File: partnerwork/design/body.py
"""One dancer: a circle, with a small chevron at its centre for the facing.

The boundary is one polar function, `outline_r`, shared by the drawing, the
hands and the routing -- so "on the border" is true by construction rather than
by two pieces of code agreeing.  It is drawn quiet and whole, broken only where
a hand mark sits on it, because how far an arm has been carried round is said
by the connection wrapping the body, not by a second arc filling up around it.
Hands sit on the rim, each in its own side's colour, the lead's a shade deeper
than the follow's.
"""
import math

from .geometry import n, polar, xy
from .style import CAP, DEEP, FAINT, INK, QUIET


BODY_R = 20               # a dancer, seen from above; their hands sit on it

RIM_W = 2.2               # one width for the whole boundary

CHEV_OUT = 7              # how far the centred chevron reaches forward
CHEV_BACK = 1             # and how little it reaches back
CHEV_HALF = 5             # half its width, well inside the rim

RIM_STEP = 3              # degrees between samples when a route walks the rim

ARM_REST = 90             # a hand at rest is a quarter of the rim from the front

R = 6                     # a hand

CAPTION_R = BODY_R + R + 2   # just past the hand it names

FREE = 0.5                # how far a hand nobody holds fades, keeping its hue

SLOT_OFFSET = 44          # how far round the rim `behind` sits from a side.
                          # Wide enough that a mark and the grey ghost of the
                          # place it left never touch, narrow enough that the
                          # spot still reads as belonging to its own side

SLOTS = ("default", "behind")

def slot_bearing(side, slot):
    """Where one of the four spots sits, as a bearing off the body's facing.

    Two sides, and on each of them the place where the arm hangs and one a
    little further round towards the back.  Off the *facing*, not off the page:
    turn a dancer and their spots turn with them.
    """
    base = -ARM_REST if side == "L" else ARM_REST
    back = -SLOT_OFFSET if side == "L" else SLOT_OFFSET
    return base + (back if slot == "behind" else 0.0)

# Where a hand settles.  `behind` carries both locks and wraps -- what tells
# them apart is which *side* the hand is behind, and what tells high from low
# is the fill the level already draws.  Nothing is invented to distinguish
# them.  A hold that names no level, or names one without saying whether it
# locks or wraps, leaves the hand where the arm hangs: there is no knowing.
SETTLE = {
    "lock": ("own", "behind"),
    "wrap": ("other", "behind"),
}

def slot_of(side, level=None, way=None):
    """The one of four spots this hand settles in: whose side, and how far
    round.  The level is not consulted -- it is a fill, not a place."""
    where, slot = (SETTLE[way] if level is not None and way in SETTLE
                   else ("own", "default"))
    return (side if where == "own" else other(side)), slot

def hand_bearing(facing, side, wind=0.0):
    return facing + (-1 if side == "L" else 1) * (ARM_REST + wind)

def settled_wind(side, level=None, way=None):
    """The winding that puts this hand in its slot.

    Winding is measured off the hand's own side and runs towards the back for
    either hand, so this is the one place the two conventions are reconciled.
    """
    aim = slot_bearing(*slot_of(side, level, way))
    return (-ARM_REST - aim) if side == "L" else (aim - ARM_REST)

def hand_point(centre, facing, side, wind=0.0):
    """Where one hand is: round the rim from the front, by however far the arm
    has carried it."""
    return polar(centre[0], centre[1], BODY_R,
                 hand_bearing(facing, side, wind))

HAND_GAP = math.degrees(math.asin((R + CAP) / BODY_R))
    # the rim's clearance around a hand mark: the same reach the connection
    # keeps, turned into arc, so the boundary and the reach stop at one border


def outline_r(delta):
    """How far the boundary is from the centre, at this bearing off the front.

    One function, used by the drawing and by anything that has to stay outside
    a body -- so `on the border` is true by construction instead of by two
    pieces of code agreeing.  A body is a plain circle now; the function stays
    because the routing reads the boundary through it.
    """
    return BODY_R

def outline_point(centre, facing, theta):
    return polar(centre[0], centre[1], outline_r(theta - facing), theta)

def rim(centre, facing, a, b, width=RIM_W):
    """One stretch of the boundary, drawn once and by one owner."""
    span = b - a
    start = outline_point(centre, facing, a)
    end = outline_point(centre, facing, b)
    large = 1 if abs(span) > 180 else 0
    sweep = 1 if span > 0 else 0
    d = f"M{xy(start)} A{BODY_R} {BODY_R} 0 {large} {sweep} {xy(end)}"
    return (f'<path d="{d}" fill="none" stroke="{QUIET}" stroke-width="{width}"'
            ' stroke-linecap="round" stroke-linejoin="round"/>')


def chevron(centre, facing):
    """The facing, said small and at the centre of the body.

    In the middle rather than on the rim, because the rim breaks for the hands
    and carries nothing else.  The centre is the one part of a dancer nothing
    else uses.
    """
    rad = math.radians(facing)
    fwd = (math.sin(rad), -math.cos(rad))
    across = (math.cos(rad), math.sin(rad))
    apex = (centre[0] + fwd[0] * CHEV_OUT, centre[1] + fwd[1] * CHEV_OUT)
    a = (centre[0] - fwd[0] * CHEV_BACK - across[0] * CHEV_HALF,
         centre[1] - fwd[1] * CHEV_BACK - across[1] * CHEV_HALF)
    b = (centre[0] - fwd[0] * CHEV_BACK + across[0] * CHEV_HALF,
         centre[1] - fwd[1] * CHEV_BACK + across[1] * CHEV_HALF)
    return (f'<polyline points="{n(a[0])},{n(a[1])} {n(apex[0])},{n(apex[1])}'
            f' {n(b[0])},{n(b[1])}" fill="none" stroke="{QUIET}"'
            ' stroke-width="1.6" stroke-linecap="round"'
            ' stroke-linejoin="round"/>')

def border(pose, who):
    """A dancer's whole boundary: one quiet outline, broken at the hands.

    It says nothing but *here is a body*.  The rim used to fill up in an arm's
    colour as that arm wound round, which was a second progress ring saying
    what the connection already says by wrapping -- so the ring is gone and the
    line keeps the job.
    """
    centre, facing = pose[who], pose[f"{who}_facing"]
    wind = pose[f"{who}_wind"]

    # Every stretch stops a hand-gap short of a hand, so the boundary never
    # runs through a mark -- and a stretch that extreme winding has squeezed
    # away is simply not drawn.
    right, left = ARM_REST + wind["R"], ARM_REST + wind["L"]
    stretches = [
        (right + HAND_GAP, 360 - left - HAND_GAP),        # behind
        (360 - left + HAND_GAP, 360 + right - HAND_GAP),  # across the front
    ]
    return "".join(rim(centre, facing, facing + a, facing + b)
                   for a, b in stretches if b - a > 0.01)

def fill_of(level, arm, deep=False):
    """The one place a level becomes a fill, so hands and pips cannot drift."""
    if level == "low":
        return DEEP[arm] if deep else INK[arm]
    if level == "above":
        return f"url(#h{arm}{'d' if deep else ''})"
    return "none"

def hand(cx, cy, leads, side, held=True, level=None, free="fade"):
    """One hand, in its own side's ink: the lead's deep, the follow's plain."""
    ink = DEEP[side] if leads else INK[side]
    stroke = ink if (held or free == "fade") else QUIET
    fill = fill_of(level, side, leads) if held else "none"
    faded = "" if held or free != "fade" else f' opacity="{FREE}"'
    dot = ""
    if level == "high":
        dot = f'<circle cx="{n(cx)}" cy="{n(cy)}" r="2.7" fill="{stroke}"/>'
    style = f"fill: {fill}; stroke: {stroke}; stroke-width: 1.5"
    if leads:
        shape = (f'<rect x="{n(cx - R)}" y="{n(cy - R)}" width="{2 * R}"'
                 f' height="{2 * R}" rx="1.5" style="{style}"{faded}/>')
    else:
        shape = (f'<circle cx="{n(cx)}" cy="{n(cy)}" r="{R}" style="{style}"'
                 f'{faded}/>')
    return shape + dot

def ring_of(pose):
    """The orbit, drawn only while one is happening.

    Nothing else in the picture is dashed, so a dashed circle says one thing:
    somebody is going round somebody.  It is centred on whoever is standing
    still -- a partner, or the midpoint when both of them travel.
    """
    if pose["ring"] is None:
        return ""
    (cx, cy), radius = pose["ring"]
    return (f'<circle cx="{n(cx)}" cy="{n(cy)}" r="{n(radius)}" fill="none"'
            f' stroke="{QUIET}" stroke-width="1" stroke-dasharray="3 4"/>')

def caption(centre, facing, side, text, wind=0.0):
    """A hand's name, set just past it and growing outwards."""
    x, y = polar(centre[0], centre[1], CAPTION_R,
                 hand_bearing(facing, side, wind))
    dx = x - centre[0]
    if dx < -2:
        anchor, dy = "end", 3
    elif dx > 2:
        anchor, dy = "start", 3
    else:
        anchor, dy = "middle", (-3 if y < centre[1] else 8)
    return (f'<text x="{n(x)}" y="{n(y + dy)}" text-anchor="{anchor}"'
            ' style="font: 8px ui-sans-serif, system-ui, sans-serif;'
            f' fill: {FAINT}">{text}</text>')

def other(side):
    return "R" if side == "L" else "L"


def side_of(where):
    """The follow's own side, from the name a hold stores their hand under."""
    return "L" if where == "left" else "R"

def hands_of(pose):
    return {who: {s: hand_point(pose[who], pose[f"{who}_facing"], s,
                                pose[f"{who}_wind"][s])
                  for s in ("L", "R")}
            for who in ("lead", "follow")}

File: partnerwork/src/partnerwork/draw/body.nim
## Draw one dancer: a circle, with a small chevron at its centre for the
## facing.
##
##   The boundary is one polar function, `outlineR`, shared by the drawing,
##     the hands and the routing -- so "on the border" is true by
##     construction rather than by two pieces of code agreeing.
##   The rim is drawn quiet and whole, broken only where a hand mark sits on
##     it: how far an arm has been carried round is said by the connection
##     wrapping the body, not by a second arc filling up around it.
##     Cost of a quiet rim: an arm carried past a full turn looks like one
##       carried less.  Accepted while nothing on the bench stores more than
##       a turn; the README keeps the question open.
##   Hands sit on the rim, each in its own side's colour, the lead's a shade
##     deeper than the follow's.
##   The six spots of rule 3 are bearings off the dancer's own facing, never
##     off the page: turn a dancer and their spots turn too.

{.experimental: "strictFuncs".}

import std/[math, options, strformat]

import ./[geometry, pose, style, terms]


const
  BODY_R* = 20.0     ## A dancer, seen from above; their hands sit on it.
  RIM_W* = 2.2       ## One width for the whole boundary.

const
  CHEV_OUT* = 7.0    ## How far the centred chevron reaches forward.
  CHEV_BACK* = 1.0   ## And how little it reaches back.
  CHEV_HALF* = 5.0   ## Half its width, well inside the rim.
  CHEV_W* = 1.6      ## The width its two legs are drawn at.

const
  RIM_STEP* = 3.0    ## Degrees between samples when a route walks the rim.
  ARM_REST* = 90.0   ## A resting hand, a quarter of the rim from the front.
  R* = 6.0           ## A hand mark's radius, or half its side.
  CAPTION_R* = BODY_R + R + 2   ## Just past the hand a caption names.
  FREE_FADE* = 0.5   ## How far a hand nobody holds fades, keeping its hue.

const SLOT_OFFSET* = 44.0
  ## How far round the rim `front` and `back` sit from a side.
  ##   Wide enough that no two marks ever touch -- two need 34.9 degrees on
  ##     this rim -- and narrow enough that a spot still belongs to its own
  ##     side.  A drawn convention, not something the dance says; the README
  ##     keeps it on the open list.

const HAND_GAP* = radToDeg(arcsin((R + CAP) / BODY_R))
  ## The rim's clearance around a hand mark: the same reach the connection
  ## keeps, turned into arc, so boundary and reach stop at one border.

const
  MARK_STROKE* = 1.5   ## The width a hand mark is outlined at.
  SEEN_GAP* = 1.2      ## Plain daylight between a reach and what it clears.
  CHEVRON_STEPS* = 6   ## Discs along one leg of a chevron, so the V is kept
                       ## clear as the shape it is.

const
  LEAD_CLEAR* = R * sqrt(2.0) + MARK_STROKE / 2 + CAP + SEEN_GAP
    ## How far a settled reach stays off a lead's hand: their mark is a
    ## square, so its corner is the far part of it (rule 22).
  FOLLOW_CLEAR* = R + MARK_STROKE / 2 + CAP + SEEN_GAP
    ## And off a follow's, whose mark is a circle.
  CHEVRON_CLEAR* = CHEV_W / 2 + CAP + SEEN_GAP
    ## And off a chevron's stroke.
    ##   The chevron is kept clear along its own two legs rather than as one
    ##     disc over the whole of it: it is a thin V pointing forward, and a
    ##     disc that covered its apex would swallow the middle of a body and
    ##     send every reach right round the outside -- which is the wrap
    ##     rule 14 forbids.  So the shape is cleared as it is drawn.


type Free* {.pure.} = enum ## Say how a hand nobody holds is drawn.
  Fade,                    ## Half strength, keeping its hue: a free hand.
  Grey                     ## Quiet outline: the ghost of a place a hand left.


#[ The Six Spots ]#

func slotBearing*(arm: Arm; slot: Slot): float =
  ## Get where one of the six spots sits, as a bearing off the body's facing.
  ##   Two sides, and on each the place where the arm hangs, one spot a
  ##     little towards the dancer's front and one a little towards their
  ##     back -- rule 3's six, named for what they are.
  let
    base = if arm == Arm.L: -ARM_REST else: ARM_REST
    forward = if arm == Arm.L: SLOT_OFFSET else: -SLOT_OFFSET
  base + (case slot
          of Slot.Front: forward
          of Slot.Back: -forward
          of Slot.Default: 0.0)


func slotOf*(arm: Arm; level: Option[Level]; way: Option[Way]):
    tuple[arm: Arm, slot: Slot] =
  ## Get the one of six spots this hand settles in: whose side, how far round.
  ##   `above` never settles anywhere but its own side's default (rule 8).
  let settled = settleOf(level, way)
  if settled.isNone:
    return (arm, Slot.Default)
  ((if settled.get.whose == Whose.Own: arm else: other(arm)),
   settled.get.slot)


func roundOf*(level: Option[Level]; way: Option[Way]): Option[Sends] =
  ## Get which way round the body this hold sends its line, where rules 4 to
  ## 6 say anything; none where nothing is said and the short way is taken.
  let settled = settleOf(level, way)
  if settled.isSome: some settled.get.sends else: none(Sends)


func handBearing*(facing: float; arm: Arm; wind = 0.0): float =
  ## Get the bearing a hand sits at: round from the front, past the rest spot
  ## by however far the arm has been carried.
  facing + (if arm == Arm.L: -1.0 else: 1.0) * (ARM_REST + wind)


func settledWind*(arm: Arm; level: Option[Level]; way: Option[Way]): float =
  ## Get the winding that puts this hand in its slot.
  ##   Winding is measured off the hand's own side and runs towards the back
  ##     for either hand, so this is the one place the two conventions are
  ##     reconciled.
  let
    landed = slotOf(arm, level, way)
    aim = slotBearing(landed.arm, landed.slot)
  if arm == Arm.L: -ARM_REST - aim else: aim - ARM_REST


func handPoint*(centre: Point; facing: float; arm: Arm; wind = 0.0): Point =
  ## Get where one hand is on the rim.
  polar(centre.x, centre.y, BODY_R, handBearing(facing, arm, wind))


func handsOf*(pose: Pose): array[Dancer, array[Arm, Point]] =
  ## Get where all four hands are.
  for who in Dancer:
    for arm in Arm:
      result[who][arm] = handPoint(
        pose.place[who], pose.facing[who], arm, pose.wind[who][arm])


#[ The Boundary ]#

func outlineR*(delta: float): float =
  ## Get how far the boundary is from the centre at this bearing off the
  ## front.
  ##   One function, used by the drawing and by anything that has to stay
  ##     outside a body -- so "on the border" is true by construction.  A
  ##     body is a plain circle now; the function stays because the routing
  ##     reads the boundary through it.
  BODY_R


func outlinePoint*(centre: Point; facing, theta: float): Point =
  ## Get the boundary point at a world bearing.
  polar(centre.x, centre.y, outlineR(theta - facing), theta)


func rim*(centre: Point; facing, a, b: float; width = RIM_W): string =
  ## Draw one stretch of the boundary, once and by one owner.
  let
    span = b - a
    start = outlinePoint(centre, facing, a)
    stop = outlinePoint(centre, facing, b)
    large = if abs(span) > 180: 1 else: 0
    sweep = if span > 0: 1 else: 0
    d = &"M{xy(start)} A{n(BODY_R)} {n(BODY_R)} 0 {large} {sweep} {xy(stop)}"
  &"""<path d="{d}" fill="none" stroke="{QUIET}" stroke-width="{width}"""" &
    " stroke-linecap=\"round\" stroke-linejoin=\"round\"/>"


func chevronPoints*(centre: Point; facing: float): array[3, Point] =
  ## Get the three points a chevron is drawn through: a wing, the apex, the
  ## other wing.
  ##   One source for the shape, so the drawing and anything that has to
  ##     keep off it read the same V (rule 22).
  let
    rad = degToRad(facing)
    fwd = (x: sin(rad), y: -cos(rad))
    across = (x: cos(rad), y: sin(rad))
  [(centre.x - fwd.x * CHEV_BACK - across.x * CHEV_HALF,
    centre.y - fwd.y * CHEV_BACK - across.y * CHEV_HALF),
   (centre.x + fwd.x * CHEV_OUT, centre.y + fwd.y * CHEV_OUT),
   (centre.x - fwd.x * CHEV_BACK + across.x * CHEV_HALF,
    centre.y - fwd.y * CHEV_BACK + across.y * CHEV_HALF)]


func chevron*(centre: Point; facing: float): string =
  ## Say the facing, small and at the centre of the body.
  ##   In the middle rather than on the rim, because the rim breaks for the
  ##     hands and carries nothing else.  The centre is the one part of a
  ##     dancer nothing else uses.
  let
    drawn = chevronPoints(centre, facing)
    (a, apex, b) = (drawn[0], drawn[1], drawn[2])
  &"""<polyline points="{n(a.x)},{n(a.y)} {n(apex.x)},{n(apex.y)}""" &
    &""" {n(b.x)},{n(b.y)}" fill="none" stroke="{QUIET}"""" &
    &" stroke-width=\"{n(CHEV_W)}\" stroke-linecap=\"round\"" &
    " stroke-linejoin=\"round\"/>"


func border*(pose: Pose; who: Dancer): string =
  ## Draw a dancer's whole boundary: one quiet outline, broken at the hands.
  ##   It says nothing but *here is a body*.  The rim used to fill up in an
  ##     arm's colour as that arm wound round -- a second progress ring
  ##     saying what the connection already says by wrapping -- so the ring
  ##     is gone and the line keeps the job.
  let
    centre = pose.place[who]
    facing = pose.facing[who]
    wind = pose.wind[who]
    right = ARM_REST + wind[Arm.R]
    left = ARM_REST + wind[Arm.L]
    # Every stretch stops a hand-gap short of a hand, so the boundary never
    # runs through a mark -- and a stretch that extreme winding has squeezed
    # away is simply not drawn.
    stretches = [
      (right + HAND_GAP, 360 - left - HAND_GAP),        # behind
      (360 - left + HAND_GAP, 360 + right - HAND_GAP),  # across the front
    ]
  for (a, b) in stretches:
    if b - a > 0.01:
      result.add rim(centre, facing, facing + a, facing + b)


#[ Hands and Furniture ]#

func fillOf*(level: Option[Level]; arm: Arm; deep = false): string =
  ## Get the fill a level draws as -- the one place a level becomes a fill,
  ## so hands and pips cannot drift.
  if level == some(Level.Low):
    return if deep: DEEP[arm] else: INK[arm]
  if level == some(Level.Above):
    return &"url(#h{arm}{(if deep: \"d\" else: \"\")})"
  "none"


func hand*(cx, cy: float; leads: bool; arm: Arm; held = true;
    level = none(Level); free = Free.Fade): string =
  ## Draw one hand, in its own side's ink: the lead's deep, the follow's
  ## plain.
  let
    ink = if leads: DEEP[arm] else: INK[arm]
    stroke = if held or free == Free.Fade: ink else: QUIET
    fill = if held: fillOf(level, arm, leads) else: "none"
    faded = if held or free != Free.Fade: ""
            else: &" opacity=\"{FREE_FADE}\""
    dot = if level == some(Level.High):
            &"""<circle cx="{n(cx)}" cy="{n(cy)}" r="2.7" fill="{stroke}"/>"""
          else: ""
    style = &"fill: {fill}; stroke: {stroke}; stroke-width: 1.5"
    shape =
      if leads:
        &"""<rect x="{n(cx - R)}" y="{n(cy - R)}" width="{n(2 * R)}"""" &
          &""" height="{n(2 * R)}" rx="1.5" style="{style}"{faded}/>"""
      else:
        &"""<circle cx="{n(cx)}" cy="{n(cy)}" r="{n(R)}" style="{style}"""" &
          &"{faded}/>"
  shape & dot


func ringOf*(pose: Pose): string =
  ## Draw the orbit, only while one is happening.
  ##   Nothing else in the picture is dashed, so a dashed circle says one
  ##     thing: somebody is going round somebody.  It is centred on whoever
  ##     is standing still -- a partner, or the midpoint when both travel.
  if pose.ring.isNone:
    return ""
  let (centre, radius) = pose.ring.get
  &"""<circle cx="{n(centre.x)}" cy="{n(centre.y)}" r="{n(radius)}"""" &
    &""" fill="none" stroke="{QUIET}" stroke-width="1"""" &
    """ stroke-dasharray="3 4"/>"""


func caption*(centre: Point; facing: float; arm: Arm; text: string;
    wind = 0.0): string =
  ## Set a hand's name just past it, growing outwards.
  let
    p = polar(centre.x, centre.y, CAPTION_R, handBearing(facing, arm, wind))
    dx = p.x - centre.x
    (anchor, dy) =
      if dx < -2: ("end", 3.0)
      elif dx > 2: ("start", 3.0)
      else: ("middle", if p.y < centre.y: -3.0 else: 8.0)
  &"""<text x="{n(p.x)}" y="{n(p.y + dy)}" text-anchor="{anchor}"""" &
    " style=\"font: 8px ui-sans-serif, system-ui, sans-serif;" &
    &""" fill: {FAINT}">{text}</text>"""
